sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala

```diff
@@ -696,6 +700,99 @@ class CudfRegexTranspiler(mode: RegexMode) {
     }
   }
 
+  private def transpile(regex: RegexAST, replacement: Option[RegexReplacement],
+      previous: Option[RegexAST]): RegexAST = {
+
+    def isMaybeEndAnchor(regex: RegexAST): Boolean = {
+      contains(regex, {
+        case RegexChar('$') | RegexEscaped('z') | RegexEscaped('Z') => true
+        case _ => false
+      })
+    }
+
+    def isMaybeNewline(regex: RegexAST): Boolean = {
+      contains(regex, {
+        case RegexChar('\r') | RegexEscaped('r') => true
+        case RegexChar('\n') | RegexEscaped('n') => true
+        case RegexChar('\f') | RegexEscaped('f') => true
+        case RegexEscaped('s') => true
+        case RegexEscaped('W') | RegexEscaped('D') =>
+          // these would get transpiled to negated character classes
+          // that include newlines
+          true
+        case RegexCharacterClass(true, _) => true
+        //TODO others?
+        case _ => false
+      })
+    }
+
+    def isMaybeEmpty(regex: RegexAST): Boolean = {
+      contains(regex, {
+        case RegexRepetition(_, term) => term match {
+          case SimpleQuantifier('*') | SimpleQuantifier('?') => true
+          case QuantifierFixedLength(0) => true
+          case QuantifierVariableLength(0, _) => true
+          case _ => false
+        }
+        case _ => false
+      })
+    }
+
+    def checkUnsupported(r1: RegexAST, r2: RegexAST): Unit = {
+//      println(s"checkUnsupported $r1 and $r2")
+      if (isMaybeEndAnchor(r1)) {
+//        println("r1 is anchor")
+        val a = isMaybeEmpty(r2)
+        val b = isMaybeNewline(r2)
+        val c = isMaybeEndAnchor(r2)
+//        println(s"r2 isMaybeEmpty: $a, isMaybeNewline, isMaybeEndAnchor: $c")
+        if (a || b || c) {
+          //TODO we should throw a more specific error
+          throw new RegexUnsupportedException(
+            "End of line/string anchor is not supported in this context")
+        }
+      }
+      if (isMaybeEndAnchor(r2)) {
+//        println("r2 is anchor")
+        val a = isMaybeEmpty(r1)
+        val b = isMaybeNewline(r1)
+        val c = isMaybeEndAnchor(r1)
+//        println(s"r1 isMaybeEmpty: $a, isMaybeNewline: $b, isMaybeEndAnchor: $c")
+        if (a || b || c) {
+          //TODO we should throw a more specific error
+          throw new RegexUnsupportedException(
+            "End of line/string anchor is not supported in this context")
+        }
+      }
+//      println("everything is fine")
+    }
+
+    def checkEndAnchorNearNewline(regex: RegexAST): Unit = {
+      regex match {
+        case RegexSequence(parts) =>
+          parts.indices.foreach { i =>
+            if (i > 0) {
```

Comment:
Is this case necessary since we are already checking both `r1` and `r2` for if either are line anchors?

Reply:
Good point